Add copy helper methods for accesses

hanslovsky opened the pull request:
This is useful for me in the imglyb context, copying from contiguous `numpy.ndarray` to imglib2 `ArrayDataAccess`.

ctrueden commented on src/main/java/net/imglib2/img/basictypeaccess/Accesses.java:
```diff
@@ -0,0 +1,136 @@
+		copy( source, target, 0, size );
+	}
+
+	public static void copy( final ByteAccess source, final ByteAccess target, final int start, final int stop )
```
This signature assumes the offset into `source` and `target` will be the same, which may not necessarily hold. I suggest modeling the API after `System.arraycopy`, which supports separate offsets for the source and dest arrays. So in this case: `copy(ByteAccess src, int srcPos, ByteAccess dest, int destPos, int length)`.

hanslovsky replied:
Good point. I wonder how complext this should be (you could even think of a different stride). Maybe it's best to just have `copy( source, target, size )`, then.

ctrueden replied:
My preference would be to match the `System.arraycopy` method. It will be familiar to Java programmers, and is more powerful.

tpietzsch replied:
would be my preference too

hanslovsky replied:
Will mimic the `System.arraycopy` method then.

ctrueden commented on src/main/java/net/imglib2/img/basictypeaccess/Accesses.java:
```diff
@@ -0,0 +1,136 @@
+	{
+		for ( int index = start; index < stop; ++index )
+		{
+			target.setValue( index, source.getValue( index ) );
```
In cases where we have an `ArrayDataAccess`, we can do better performance-wise, by calling `getCurrentStorageArray()` followed by `System.arraycopy`. I suggest adding this as a special case somehow.

hanslovsky replied:
I wonder if this is pre-mature optimization. If copying from one access to another access is a bottleneck, there is probably something wrong with the code. Also, `ArrayDataAccess.getCurrentStorageArray()` returns `Object` and is not even guaranteed to return array.
 A way around this would be to add
```java
public default void copyInto( ByteAccess target )
{
// do the copy as in the proposed methods
}
```
to the `ByteAccess` etc. interfaces and override with special case implementations for `ByteArray` etc.

hanslovsky replied:
FWIW, I made a little "benchmark":
```java
import net.imglib2.img.basictypeaccess.ByteAccess;
import net.imglib2.img.basictypeaccess.array.ByteArray;

public class ByteAccessCopyTest
{

	public static void main( final String args[] )
	{
		final int iterations = 100000;

		final int size = 1000000;

		final ByteArray source = new ByteArray( size );
		final ByteArray target = new ByteArray( size );

		for ( int i = 0; i < iterations; ++i )
		{
			copy( source, target, size );
		}

		final long tStartAccessCopy = System.nanoTime();
		for ( int i = 0; i < iterations; ++i )
		{
			copy( source, target, size );
		}
		final long tStopAccessCopy = System.nanoTime();
		final long tAccessCopy = tStopAccessCopy - tStartAccessCopy;

		for ( int i = 0; i < iterations; ++i )
		{
			System.arraycopy( source.getCurrentStorageArray(), 0, target.getCurrentStorageArray(), 0, iterations );
		}

		final long tStartArraycopy = System.nanoTime();
		for ( int i = 0; i < iterations; ++i )
		{
			System.arraycopy( source.getCurrentStorageArray(), 0, target.getCurrentStorageArray(), 0, iterations );
		}
		final long tStopArraycopy = System.nanoTime();
		final long tArraycopy = tStopArraycopy - tStartArraycopy;

		System.out.println( "access copy: " + tAccessCopy * 1.0 / iterations );
		System.out.println( "array  copy: " + tArraycopy * 1.0 / iterations );

	}

	private static final void copy( final ByteAccess source, final ByteAccess target, final int size )
	{
		for ( int i = 0; i < size; ++i )
		{
			target.setValue( i, source.getValue( i ) );
		}
	}

}
```

```
access copy: 111722.59448
array  copy: 4472.66502
```
Might be worth having those special case implementations. Is there a good way to check if an `Object` is instance of primtiive type other than doing `obj instance of byte[] || obj instance of char[]` etc.?

ctrueden replied:
I expanded your performance test to include `ShortArray` and `FloatArray` as well mixed into one scope, and speed remains equivalent to `System.arraycopy`. Pretty impressive JIT!

hanslovsky commented:
I updated the interface to follow `System.arraycopy`. I c&p'ed and modified the relevant parts of the `System.arraycopy` documentation. Is that an issue?

hanslovsky commented:
Would it be useful to add a method
```java
copy( Object src, int, Object dest, int, int )
```
that picks the right `copy` method or throws if `src` and `dest` are not the same kind of access?

tpietzsch commented on src/main/java/net/imglib2/img/basictypeaccess/Accesses.java:
```diff
@@ -0,0 +1,451 @@
+		if ( src == dest )
+		{
+			if ( srcPos == destPos ) { return; }
+			if ( srcPos < destPos )
```
This looks incomplete. What if `destPos > srcPos`

hanslovsky replied:
Addressed in e4897a5

hanslovsky commented on src/main/java/net/imglib2/img/basictypeaccess/Accesses.java:
```diff
@@ -0,0 +1,451 @@
+					dest.setValue( index + destPos, src.getValue( index + srcPos ) );
+				}
+			}
+			return;
```
Good catch, the `return` should be within the inner `if` block. This should resolve this comment and the other one, as well. Will add appropriate test cases to make sure all three cases are handled.

maarzt commented:
The PR introduces a lot of code duplication. Which could by completely removed.
One way would to add an `CopyElement` interface to all the accesses.
```
interface CopyElement<A> {
   public copyElement(int srcIndex, A destAccess, int destIndex);
}
```
Accesses.copy would only be needed once for this CopyElement class

hanslovsky commented:
Then the code duplication would reside within each of the basic access interfaces (`ByteAccess`, `CharAccess`, etc.) instead of a single class because there are no c++ like templates in Java, and on top of that we would have introduced another interface. I don't see any benefits in that.

maarzt commented:
Yes, Java generics are barely usable at all that's true :rage:. But that's no excuse :angel:. Despite all the pain, super cool stuff like for example imglib2 can be done :heart:. That reminds me, `ByteType` enables fast access to the elements of `ByteAccess`. And similar stuff is possible for `IntAcess`, `ShortAccess`... We could use that, to kill all the code duplication:
https://gist.github.com/maarzt/77ee5ad4de5a380ad95daf4e3bbae9ca

maarzt commented:
It would be totally useful to have an interface, maybe called `BasicAccess`, that's implemented by all accesses. I find it very wired, that `ByteAccess` and `IntAccess` share no common interface.
The `BasicAccess` interface could have such insanely useful methods like `public int size()`.

And one could write generic methods like:
```
public <A extends BasicAccess> void copy(A source, int sourceIndex, A destination, int destIndex, int size);
```

hanslovsky commented:
The code in your gist looks interesting and would certainly reduce code duplication. What if `src` and `dest` are different implementations of the same interface, e.g. `SomeByteAccess src` and `SomeOtherByteAccess dest` -- would that still compile?

As for a common interface for all accesses, that is probably not within the scope of this PR.

hanslovsky commented:
I implemented a more generic interface to reduce code duplication as suggested by @maarzt 
This comes at the cost of:
 - support for `CharAccess`
 - compile time safety as now any `Object` can be passed for `src` and `dest` (exception is thrown if incompatible combination of `src`, `dest` is passed)

hanslovsky commented:
What are the opinions on this?

> This comes at the cost of:
> - support for CharAccess
> - compile time safety as now any Object can be passed for src and dest (exception is thrown if incompatible combination of src, dest is passed)

StephanPreibisch commented:
Hi, while this is useful I was wondering if this should be in core ... maybe we should have something like imglib2-utils? What do you think?

Some things from imglib2/src/main/java/net/imglib2/util/ could also go there ... including quite some copy methods in there.

Cheers,
Stephan

maarzt commented:
I would suggest to put it into `imglib2-algorithms`

hanslovsky commented:
I disagree. I think that imglib2 core is a good place for something as simple as a copy and as closely tied to the access interfaces. Of course, we could modularize more, but I think that, if we go that route, we will find many things that could go into their own separate modules, e.g.
 - imglib2-convert
 - imglib2-views
 - imglib2-transform
 - imglib2-awt

That's mostly a matter of taste.

imagejan commented:
> That's mostly a matter of taste.

It's also a matter of dependency hierarchy, no? Would any consumers of `imglib2` benefit of a more modular component setup? Are there (many) projects that would require `imglib2` but _not_ `imglib2-algorithm` (or any of the potential sub-projects, for that matter) and therefore benefit from a smaller footprint?

maarzt commented:
To me it feels like: This will lay around, convolute the code, and not being used much. An imglib2 user is not intended to do sophisticated actions on Accesses. Imglyb is a very specific (useful) usecase.
@hanslovsky What am I missing? Why do you think others would need it as part of imglib2 core?

hanslovsky commented:
> An imglib2 user is not intended to do sophisticated actions on Accesses. 

That is just plain wrong. There is different levels of expertise, and some will deal with accesses directly.

There is no algorithmic element to this, so no reason to put this into imglib2-algorithm.

maarzt commented:
StephanPreibisch wrote:
> ... maybe we should have something like imglib2-utils? What do you think?

If imglib2 has such a diverse set of users (beginners that script in FIJI to n5 and imglyb). It's maybe realy beneficial if there is imglib2-expert-utils, such that a less experienced user isn't confused by all the wild range of functionality. Or imglib2-backend-utils?

axtimwalde commented:
@maazt @hanslovsky I agree with @hanslovsky that this is a core/ utils component because it does not introduce dependencies, has no side effects, and enables core to be used in a way that is necessary.  On the other hand, I do not see any use for it outside of imglyb currently, and it does not require changes in core for it to work, so I would keep this code in imglyb until the need arises to use it elsewhere.  This will also give it time to generalize/ modularize better and possibly support other types.  We have been following this pattern elsewhere, e.g. BDV.  @maarzt I do not share the sentiment that this code would "lay around and convolute the code" because it does not intersect with other aspects of core.

axtimwalde commented:
@imagejan I am using imglib2 without algorithm all the time but would not benefit from excluding stuff from the existing imglib2 core which is rather small.  Our previous approach about modularization was along dependency lines.  This is the only reason why imglib2-realtransform is not in imglib2 core.

hanslovsky commented:
The consesus seems to be that this adds not enough usabiltiy to imglib2 core, gonna close.

hanslovsky commented:
For now, people can use imglib2-imglyb to use `Accesses.copy`:
https://github.com/imglib/imglib2-imglyb/blob/c6dda7d07808e860e2f727cb2708fa1a5878462f/src/main/java/tmp/net/imglib2/img/basictypeaccess/Accesses.java
Right now I do not see a good way to create a static method that is specialized for each access type (array access, unsafe access, etc.) without adding interface (methods). We should probably re-visit if there is ever a need outside imglyb.
